### PR TITLE
Update Katib on SQLFlow desgin

### PR DESCRIPTION
**Pull request by samplise**

### CODE REVIEW

File: doc/design/katib_on_sqlflow.md [hunk 1]
```diff
@@ -64,28 +64,56 @@ We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_co
...
+- `image` specifies the container image source for the Katib tuning job.
+- `sql` sql statement input by users.
+
+## Run Tuning Job on Katib
+
+In each Katib tuning job, users need to define tuning parameters (i.e., the hyperparameter's name, type and range) in a model at first. Then Katib will start multiple parallel Pods, which will take different value of those hyperparameters, to train and measure the model.
+
+For SQLFlow, each Katib pod will execute the following command:
+
+`repl -m "select * from c to train ... "; python katib_xgb_submitter.py ...`
```
Comment (wangkuiyi) [inline doc/design/katib_on_sqlflow.md]: Do you want to replace `;` with `|`?

Comment (wangkuiyi) [reply doc/design/katib_on_sqlflow.md]: How comes the file `katib_xgb_submitter.py`?

Comment (samplise) [reply doc/design/katib_on_sqlflow.md]: This is a design issue that I hope to have a discussion here.  I am trying to find a way to pass the `IR` of the input sql query to each Katib pod. 

One is to use sqlflow generate the submitter in the Katib pod when it is started, which is something like `katib_xgb_submitter.py`. Then we execute this python program in the pod and train the model.  So the command to generate submitter program should be something like `sqlflow "select * from ..."` ? 

The second way is to ` repl -m "select * from ..." | python sqlflow_submitter/couler/katib/train.py  `. However, we should be careful here since Katib will append tuning hyperparameters to the command line. The final command will be `repl -m "select * from ..." | python sqlflow_submitter/couler/katib/train.py --num_round 60 --max_depth 3`. We need to make sure `num_round` and `max_depth` are passed to the train.py correctly.

File: doc/design/katib_on_sqlflow.md [hunk 2]
```diff
@@ -64,28 +65,59 @@ We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_co
...
+- `sql` sql statement input by users.
+- `datasource` train and validation data source 
+
+## Run Tuning Job on Katib
+
+In each Katib tuning job, users need to define tuning parameters (i.e., the hyperparameter's name, type and range) in a model at first. Then Katib will start multiple parallel Pods, which will take different value of those hyperparameters, to train and measure the model.
+
+For SQLFlow, each Katib pod will execute the following command:
+
+`repl -m "select * from c to train ... " | python sqlflow_submitter/couler/katib/xgboost_train.py ...`
```
Comment (Yancey1989) [inline doc/design/katib_on_sqlflow.md]: For my understanding, the Katib pod runs the SQL with the specified hyperparameter such as  `max_depth =3`, but the user input SQL is `max_depth =[2, 10]`, who implement this translation?

File: doc/design/katib_on_sqlflow.md [hunk 3]
```diff
@@ -64,28 +65,59 @@ We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_co
...
```
Comment (Yancey0623) [inline doc/design/katib_on_sqlflow.md]: What does `repl -m` mean? If you want to generate the IR proto message in text format, I believe it's `repl -parse`.

Comment (Yancey0623) [inline doc/design/katib_on_sqlflow.md]: `python sqlflow_submitter/couler/katib/xgboost_train.py`
=>
`python -m sqlflow_submitter.katib.xgboost.train`

File: doc/design/katib_on_sqlflow.md [hunk 4]
```diff
@@ -64,28 +65,59 @@ We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_co
...
+
+Program `xgboost_train.py` will received the value for `num_round` and `max_depth` from Katib, and read train and validation data from the data source at first. Then this program trains the model and output the measurement for this model. Based on the output, Katib will choose the next set of value for `num_round` and `max_depth` and starts new pods for training.
+
+## Pipeline
+
+The pipeline from SQL statements to Argo workflow:
+
+- SQLFlow generate `IR` from input SQL statements.
+- `couler_katib_codegen.go` take this `IR` as input and obtains parameters for Katib training job.
+- `couler_katib_codegen.go` generates a Python program which invokes `couler.sqlflow.train`. At the same time, `couler_katib_codegen.go` fills this API's arguments with Katib parameters.
```
Comment (Yancey0623) [inline doc/design/katib_on_sqlflow.md]: I think there is only one file `pkg/sql/codegen/couler/codegen.go` to generate the Couler Program. Why there are two files that can generate the Couler Program?

File: doc/design/katib_on_sqlflow.md [hunk 5]
```diff
@@ -64,28 +65,59 @@ We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_co
...
-couler.katib.train(model_def, data="/hdfs/temp")
+couler.maxcompute.run("""SELECT * FROM a, b WHERE a.id = b.id INTO c;""")
+couler.sqlflow.train(model="xgboost.gbtree", hyperparameters={"objective": "multi:softmax", 
+    "eta": 0.1},  image="data_scientist/xgboost:v0.5",
+    sql="select * from c to train data_scientist/xgboost:v0.5/xgboost.gbtree ... ",
+    datasource="mysql://..." )
 ```
 
-## `couler.katib.train(...)`
+## `couler.sqlflow.train(...)`
```
Comment (Yancey0623) [inline doc/design/katib_on_sqlflow.md]: I prefer to keep the name to `couler.katib.train`

File: doc/design/katib_on_sqlflow.md [hunk 6]
```diff
@@ -64,28 +65,59 @@ We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_co
...
 
-SQLFlow parses each SQL program into an IR, which is a list of statement IRs.  The `codegen_couler.go` converts the IR into a Couler program.   We need to add a Couler functions `couler.katib.train` for the calling by the generated Couler program.
+SQLFlow parses each SQL program into an IR, which is a list of statement IRs.  The `codegen_couler.go` converts the IR into a Couler program.   We need to add a Couler functions `couler.sqlflow.train` for the calling by the generated Couler program.
 
 Consider the following example program.
 
 ```sql
 SELECT * FROM a, b WHERE a.id = b.id INTO c;
-SELECT * FROM c TO TRAIN model_def WITH objective=multi:softmax, eta=1 LABEL class INTO my_xgb_model;
+SELECT * FROM c TO TRAIN data_scientist/xgboost:v0.5/xgboost.gbtree
```
Comment (Yancey0623) [inline doc/design/katib_on_sqlflow.md]: Maybe we can add attribute `range.max_depth=[2, 10]` in the demo so that users can know it's an auto hyperparameter search job.

File: doc/design/katib_on_sqlflow.md (hunk 5 above)
Comment (samplise) [reply doc/design/katib_on_sqlflow.md]: For katib in couler, it may also serve other applications. So we design a unique API for each application since their parameters may be quite different.

File: doc/design/katib_on_sqlflow.md (hunk 6 above)
Comment (samplise) [reply doc/design/katib_on_sqlflow.md]: It may be a way to label it is a katib training job. However, the concept of auto model training is that users need not specify the range for any parameter, we need to fill default value instead.

I still do not find a good way to label it is a katib training job in sql statement. Any suggestions?

File: doc/design/katib_on_sqlflow.md (hunk 5 above)
Comment (Yancey0623) [reply doc/design/katib_on_sqlflow.md]: `couler.sqlflow.train` seems a general module to train all sqlflow models, maybe `couler.sqlflow.katib.train` ?

File: doc/design/katib_on_sqlflow.md [hunk 7]
```diff
@@ -64,28 +65,102 @@ We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_co
...
+
+For example, users may define the following command for tuning job Pod:
+
+`python -m sqlflow_submitter.couler.katib.xgboost_train`
+
+The actual command during runtime will be:
+
+`python -m sqlflow_submitter.couler.katib.xgboost_train --max_depth 5 ...`, hyperparameter `max_depth` is added by Katib.
+
+We have proposed two different designs here. Assuming users have input the following SQL statement:
```
Comment (Yancey0623) [inline doc/design/katib_on_sqlflow.md]: Which one does this design doc choice?

File: doc/design/katib_on_sqlflow.md [hunk 8]
```diff
@@ -64,28 +65,102 @@ We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_co
...
+
+In the SQLFlow submitter, it invokes `couler.sqlflow.train`:
+
+`couler.sqlflow.train(model="xgboost", hyperparameters={"objective": ... }, params={"select": ..., "features": ...}, datasource=... )`
+
+Then in tuning Pod, it executes the following command:
+
+`python -m xgb_train --select "select ..." --features "{...}" ... --max_depth 5 ...`
+
+### Pass SQL to tuning Pod and extract parameters in the Pods
```
Comment (Yancey0623) [inline doc/design/katib_on_sqlflow.md]: I prefer the second one,  consider the following SQL program:
``` 
create table tmp_t as (...)
select * from tmp_t to train ...
```

The `features` field depends on the [Feature Derivation](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/design/feature_derivation.md ) which read some table data from `tmp_t`, so we should generate the IR at the step container running instead of generating the Coulr program.

File: doc/design/katib_on_sqlflow.md [hunk 9]
```diff
@@ -64,28 +65,102 @@ We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_co
...
+
+In this design, SQLFlow parses the input SQL statement but only extract tuning hyperparameters. Then SQLFlow passes the whole SQL statement to `couler.sqlflow.train`. 
+
+In the SQLFlow submitter, it invokes `couler.sqlflow.train`:
```
Comment (Yancey0623) [inline doc/design/katib_on_sqlflow.md]: > In the SQLFlow submitter, it invokes `couler.sqlflow.train`:

I'm confused about this, `SQLFlow submitter` should like tensorflow/xgboost training/prediction program, why invokes a Couler step function ?

File: doc/design/katib_on_sqlflow.md (hunk 6 above)
Comment (Yancey0623) [reply doc/design/katib_on_sqlflow.md]: Maybe we can use env to label it, like `SQLFLOW_XGBOOST_SUBMITTER=katib`

**Review by Yancey0623 — COMMENTED** [review]
Another question, how to fetch the logs from a Katib Job? The current [Fetch](https://github.com/sql-machine-learning/sqlflow/blob/develop/pkg/argo/fetch.go#L45) implementation works well on Katib Job?
This question does not block this PR merge, but maybe we should consider it before coding.

File: doc/design/katib_on_sqlflow.md (hunk 7 above)
Comment (aeonproject) [reply doc/design/katib_on_sqlflow.md]: The current implementation is following the first approach.

Comment (samplise) [reply doc/design/katib_on_sqlflow.md]: Now I am following the first approach. I am fine with the other one.

File: doc/design/katib_on_sqlflow.md (hunk 8 above)
Comment (samplise) [reply doc/design/katib_on_sqlflow.md]: Fine with me.

File: doc/design/katib_on_sqlflow.md [hunk 10]
```diff
@@ -64,28 +65,75 @@ We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_co
...
+
+In the last step, it runs following codes in `xgb_train`:
+
+```python
+setenv("MAX_DEPTH", 5)
+run_cmd("repl -m \"select * ...\" ")
+
+```
+
+## TODO
```
Comment (Yancey0623) [inline doc/design/katib_on_sqlflow.md]: The TODO list can be an issue like: https://github.com/sql-machine-learning/sqlflow/issues/1066

File: doc/design/katib_on_sqlflow.md [hunk 11]
```diff
@@ -64,28 +65,75 @@ We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_co
...
 ```
 
-## `couler.katib.train(...)`
+## `couler.sqlflow.katib.train(...)`
 
 Considering Katib itself supports multiple models and frameworks, and more may come in the future, we introduce the following Couler function.
 
 ```python
-def couler.katib.train(model_def=None, hyperparameters={})
+def couler.sqlflow.train(model=None, hyperparameters={}, image=None, sql=None,datasource=None)
```
Comment (Yancey0623) [inline doc/design/katib_on_sqlflow.md]: `couler.sqlflow.katib.train` ?

**Review by Yancey0623 — DISMISSED** [review]
LGTM with tiny comment

**Review by Yancey0623 — APPROVED** [review]
LGTM
